mm2src/coins/utxo/rpc_clients/electrum_rpc/client.rs

```diff
@@ -0,0 +1,1056 @@
+            self.clone()
+                .electrum_request_to(to_addr.clone(), request)
+                .map_ok(|response| (JsonRpcRemoteAddr(to_addr), response))
+                .boxed()
+                .compat(),
+        )
+    }
+}
+
+#[cfg_attr(test, mockable)]
+impl ElectrumClient {
+    pub async fn try_new(
+        client_settings: ElectrumClientSettings,
+        event_handlers: Vec<Box<SharableRpcTransportEventHandler>>,
+        block_headers_storage: BlockHeaderStorage,
+        abortable_system: AbortableQueue,
+        scripthash_notification_sender: Option<UnboundedSender<ScripthashNotification>>,
+    ) -> Result<ElectrumClient, String> {
+        let client = ElectrumClient(
+            ElectrumClientImpl::try_new_arc(
+                client_settings,
+                block_headers_storage,
+                abortable_system,
+                event_handlers,
+                scripthash_notification_sender,
+            )
+            .await?,
+        );
+
+        Ok(client)
+    }
+
+    /// Sends a JSONRPC request to all the connected servers.
+    ///
+    /// This method will block until a response is received from at least one server.
+    async fn electrum_request_multi(
+        self,
+        request: JsonRpcRequestEnum,
+    ) -> Result<(JsonRpcRemoteAddr, JsonRpcResponseEnum), JsonRpcErrorType> {
+        // Whether to send the request to all active connections or not.
+        let send_to_all = matches!(request, JsonRpcRequestEnum::Single(ref req) if SEND_TO_ALL_METHODS.contains(&req.method.as_str()));
+        // Request id and serialized request.
+        let req_id = request.rpc_id();
+        let request = json::to_string(&request).map_err(|e| JsonRpcErrorType::InvalidRequest(e.to_string()))?;
+        let request = (req_id, request);
+        // Use the active connections for this request.
+        let connections = self.connection_manager.get_active_connections().await;
+        // Maximum number of connections to establish or use in request concurrently.
+        // Could be up to connections.len().
+        let concurrency = 1;
+        match self
+            .send_request_using(&request, connections, send_to_all, concurrency)
+            .await
+        {
+            Ok(response) => Ok(response),
+            // If we failed the request using only the active connections, try again using all connections.
+            Err(_) => {
+                let connections = self.connection_manager.get_all_connections();
+                // The concurrency here must be `1`, because we are trying out connections that aren't maintained
+                // which means we might break the max connections rule.
+                // We will at most we will break this rule by `1` (have `max_connected + 1` open connections).
+                let concurrency = 1;
+                match self
+                    .send_request_using(&request, connections, send_to_all, concurrency)
+                    .await
```

Comment:
I thought about this but left it as is just to not make a special case for `ping` (trying not to think that `SEND_TO_ALL_METHODS` is only the ping method).

We could avoid sending the request all together here, the ping is only used to check if the server is still connected and also keeping the connection alive. If we fallback to this step then we for sure disconnected all the active servers cuz they errored, still not wanna think about it so specific though.

Reply:
something to notice here as well related to the `max_connected` threshold, if `send_to_all` is set and we are trying for all connections, this will not respect the threshold, so we better close the connections right after we send the request but only in the fallback case.